### PR TITLE
Pootle fs state and response

### PR DESCRIPTION
**Pull request by phlax**
Adds state and response objects to be used by pootle_fs, and tests

### CODE REVIEW

File: pootle/apps/pootle_fs/response.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    "description":
+        ("Merged - Pootle won where unit updates conflicted")}
+
+
+class ActionStatus(Status):
+
+    def __init__(self, action_type, original_status, complete=True, msg=None):
+        self.action_type = action_type
+        self.complete = complete
+        self.original_status = original_status
```
Comment (unho) [inline pootle/apps/pootle_fs/response.py]: Nit, assign attributes in the same order as parameters.

File: pootle/apps/pootle_fs/response.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        ("Merged - FS won where unit updates conflicted")}
+FS_ACTION["merged_from_pootle"] = {
+    "title": "Merged from pootle",
+    "description":
+        ("Merged - Pootle won where unit updates conflicted")}
+
+
+class ActionStatus(Status):
+
+    def __init__(self, action_type, original_status, complete=True, msg=None):
```
Comment (unho) [inline pootle/apps/pootle_fs/response.py]: Which is the point on accepting `msg`? Is `ActionStatus` meant to be subclassed?

File: pootle/core/state.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    @property
+    def has_changed(self):
+        return any(self.__state__.values())
+
+    def add(self, k, v):
+        if k in self.__state__:
+            self.__state__[k].append(v)
+
+    @property
+    def states(self):
```
Comment (unho) [inline pootle/core/state.py]: Please put the properties together, following convention for order in models (despite this is not a model) http://docs.translatehouse.org/projects/pootle/en/latest/developers/styleguide.html#order-in-models

File: tests/misc/state.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        def state_foo(self, **kwargs):
+            for x in [1, 2, 3]:
+                yield {str(x): x}
+
+        def state_bar(self, **kwargs):
+            for x in [4, 5, 6, 7, 8, 9]:
+                yield {str(x): x}
+
+        def state_baz(self, **kwargs):
+            yield dict(never="called")
```
Comment (unho) [inline tests/misc/state.py]: Nit: `yield {'never': "called"}`

File: tests/misc/state.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+
+@pytest.mark.django_db
+def test_state_states():
+
+    class ContextualState(State):
+
+        @property
+        def states(self):
+            return ["foo", "bar", "empty"]
```
Comment (unho) [inline tests/misc/state.py]: Nit: Use alphabetical order as below.

Comment (unho) [inline tests/misc/state.py]: nvm. I see that it is intended.

File: pootle/core/response.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+
+class ItemResponse(object):
+
+    def __init__(self, response, action_type, complete=True, msg=None, **kwargs):
+        self.response = response
+        self.action_type = action_type
+        self.complete = complete
+        self.kwargs = kwargs
+        self.msg = msg or ""
```
Comment (unho) [inline pootle/core/response.py]: Nit: Set in the same order as in arguments.

File: pootle/core/response.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+                yield k
+
+    def __len__(self):
+        return len([x for x in self.__iter__()])
+
+    def __str__(self):
+        failed = ""
+        if self.has_failed:
+            failed = "FAIL "
+        if not self.made_changes:
```
Comment (unho) [inline pootle/core/response.py]: Nit: In case like this I find cleaner to invert the order so `not` is not used.

File: pootle/core/response.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+               self.context,
+               failed,
+               self.success))
+
+    @property
+    def success(self):
+        return ', '.join(
+            ["%s: %s" % (k, len(list(self.completed(k))))
+             for k in self.response_types
+             if (len(list(self.completed(k))))])
```
Comment (unho) [inline pootle/core/response.py]: Nit: is the parentheses around `len` actually necessary. It already has a lot of them.

File: pootle/core/response.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+    @property
+    def response_types(self):
+        return self.__responses__.keys()
+
+    @property
+    def has_failed(self):
+        return len(list(self.failed())) and True or False
```
Comment (unho) [inline pootle/core/response.py]: `return len(list(self.failed())) > 0`

File: pootle/core/response.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+    @property
+    def made_changes(self):
+        for complete in self.completed():
+            return True
+        return False
```
Comment (unho) [inline pootle/core/response.py]: `return len(self.completed()) > 0`?

**Comment by unho** [comment]
lgtm nits aside.

File: tests/misc/state.py (hunk 4 above)
Comment (phlax) [inline tests/misc/state.py]: im happy to use dict(), esp in a test

there are situations when an explicit `{}` is better (ie when you need to use vars for keys), but on the whole, i prefer using the `dict` builtin

File: pootle/core/state.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    def __str__(self):
+        return (
+            "<%s(%s): %s %s>"
+            % (self.__class__.__name__,
+               self.state.context,
+               self.state_type,
+               self.kwargs))
+
+    def __getattr__(self, k):
+        if k in self.kwargs:
```
Comment (ta2-1) [inline pootle/core/state.py]: is this uncovered?

Comment (phlax) [inline pootle/core/state.py]: it gets covered in the next PR, but yep, it should be tested in its own right - ill add

File: tests/misc/state.py (hunk 5 above)
Comment (ta2-1) [inline tests/misc/state.py]: is it necessary? I think that `State.states` should have the same result.

Comment (phlax) [inline tests/misc/state.py]: no the default algo would include `baz` that is what this is testing

Comment (phlax) [inline tests/misc/state.py]: ill add a docstring

File: pootle/core/state.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+    def __iter__(self):
+        for k in self.__state__:
+            if self.__state__[k]:
+                yield k
+
+    def __str__(self):
+        if not self.has_changed:
+            return (
+                "<%s(%s): Everything up-to-date>"
```
Comment (ta2-1) [inline pootle/core/state.py]: "Everything up-to-date" is default str() and we put `State` to `pootle.core.state`. I feel that `State` is going to be used for tracking update changes but there is no infomation here about that.

**Comment by ta2-1** [comment]
lgtm

**Pull request merged**